### PR TITLE
Vision domain classifier issue 841

### PR DESCRIPTION
**Pull request by nirhutnik**
Added WholeDatasetDrift

### CODE REVIEW

File: tests/vision/vision_conftest.py [hunk 1]
```diff
@@ -48,24 +48,24 @@
 
 @pytest.fixture(scope='session')
 def mnist_data_loader_train():
-    return load_mnist_dataset(train=True, object_type='DataLoader')
+    return load_mnist_dataset(train=True, object_type='DataLoader', shuffle=False)
```
Comment (noamzbr) [inline tests/vision/vision_conftest.py:51]: can't set a seed for shuffling?

File: deepchecks/vision/checks/distribution/whole_dataset_drift.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+                            'area',
+                            'brightness',
+                            'normalized_red_mean',
+                            'normalized_green_mean',
+                            'normalized_blue_mean']
+
+
+class WholeDatasetDrift(TrainTestCheck):
+    """
+    Calculate drift between the entire train and test datasets using a model trained to distinguish between them.
```
Comment (noamzbr) [inline deepchecks/vision/checks/distribution/whole_dataset_drift.py:32]: This first line should say it's trying to distinguish based on simple image properties

File: deepchecks/vision/checks/distribution/whole_dataset_drift.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+    Check fits a new model to distinguish between train and test datasets, called a Domain Classifier.
```
Comment (noamzbr) [inline deepchecks/vision/checks/distribution/whole_dataset_drift.py:34]: Worth adding a few words of "why?" - why is this check useful for me and when should I run it,

File: deepchecks/vision/checks/distribution/whole_dataset_drift.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    properties for each image (such as brightness, aspect ratio etc.) and uses them as input features to the Domain
+    Classifier.
+    Once the Domain Classifier is fitted the check calculates the feature importance for the domain classifier
+    model. The result of the check is based on the AUC of the domain classifier model, and the check displays
+    the change in distribution between train and test for the top features according to the
+    calculated feature importance.
+
+    Parameters
+    ----------
+    alternative_image_properties : List[str] , default: None
```
Comment (noamzbr) [inline deepchecks/vision/checks/distribution/whole_dataset_drift.py:45]: No custom callables?

File: deepchecks/vision/checks/distribution/whole_dataset_drift.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        List of alternative image properties names. Must be attributes of the ImageFormatter classes that are passed to
+        train and test's VisionData class. If None, check uses DEFAULT_IMAGE_PROPERTIES.
+    n_top_columns : int , default: 3
```
Comment (noamzbr) [inline deepchecks/vision/checks/distribution/whole_dataset_drift.py:48]: All these parameters should be adjusted. E.g. features / columns -> properties, if that's how we are calling these. Also when talking about categories - categories of the image extracted properties.

**Review by noamzbr — COMMENTED** [review]
General question - I'm a user, and I want to understand what "_normalized_rgb_mean" means and does, how would I go about it?

File: tests/vision/vision_conftest.py (hunk 1 above)
Comment (nirhutnik) [reply tests/vision/vision_conftest.py:51]: The shuffling is done inside pytorch, didn't seem like I could.

File: deepchecks/vision/checks/distribution/whole_dataset_drift.py (hunk 2 above)
Comment (nirhutnik) [reply deepchecks/vision/checks/distribution/whole_dataset_drift.py:32]: Done

File: deepchecks/vision/checks/distribution/whole_dataset_drift.py (hunk 4 above)
Comment (nirhutnik) [reply deepchecks/vision/checks/distribution/whole_dataset_drift.py:45]: Thought the user can insert them into their ImageFormatter class.

**Comment by nirhutnik** [comment]
<p>Fixed the second comment,</p><p>Not sure what you want me to do with the first</p>

---

 <a href='https://app.reviewnb.com/deepchecks/deepchecks/pull/879/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

File: tests/vision/vision_conftest.py (hunk 1 above)
Comment (noamzbr) [reply tests/vision/vision_conftest.py:51]: Try this?
https://pytorch.org/docs/stable/notes/randomness.html

File: deepchecks/vision/checks/distribution/whole_dataset_drift.py (hunk 4 above)
Comment (noamzbr) [reply deepchecks/vision/checks/distribution/whole_dataset_drift.py:45]: Mmm doesn't seem comfortable to me - they need to subclass, when the only reason to subclass is to add a few methods? Passing callable seems more natural,

File: deepchecks/vision/checks/distribution/whole_dataset_drift.py (hunk 5 above)
Comment (noamzbr) [reply deepchecks/vision/checks/distribution/whole_dataset_drift.py]: Yeah I agree, but we need to note we're talking about the importance of general image properties to a domain classifier model (which is not the user model)

**Comment by noamzbr** [comment]
<p>first comment was about the fact that you say that a multivariate drift occurs when the properties of the input feature change (change in P(X)). Need to stress that multivariate drift is change in the relation between input features.</p>

---

 <a href='https://app.reviewnb.com/deepchecks/deepchecks/pull/879/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by noamzbr** [comment]
<p>mmm ok, I'm still afraid the user will stop mid-read because he'd think he stumbled on the tabular docs. </p>

---

 <a href='https://app.reviewnb.com/deepchecks/deepchecks/pull/879/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

File: deepchecks/core/check_utils/whole_dataset_drift_utils.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+
+    values_dict = {
+        'domain_classifier_auc': domain_classifier_auc,
+        'domain_classifier_drift_score': auc_to_drift_score(domain_classifier_auc),
+        'domain_classifier_feature_importance': fi.to_dict() if fi is not None else {},
+    }
+
+    headnote = f"""
+    <span>
+    The shown features are the features that are most important for the domain classifier - the
```
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/core/check_utils/whole_dataset_drift_utils.py]: Fixed

File: tests/vision/vision_conftest.py (hunk 1 above)
Comment (nirhutnik) [reply tests/vision/vision_conftest.py:51]: Thanks

File: deepchecks/vision/checks/distribution/whole_dataset_drift.py (hunk 5 above)
Comment (nirhutnik) [reply deepchecks/vision/checks/distribution/whole_dataset_drift.py]: Added description in docstring

File: deepchecks/vision/checks/distribution/whole_dataset_drift.py (hunk 3 above)
Comment (nirhutnik) [reply deepchecks/vision/checks/distribution/whole_dataset_drift.py:34]: I love it when you comment on basically your own text :)
But seriously - this is a lot of text anyway, and I don't see us "justify" when to use a check in other places, so not sure what you mean here

**Comment by nirhutnik** [comment]
<p>now?</p>

---

 <a href='https://app.reviewnb.com/deepchecks/deepchecks/pull/879/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by nirhutnik** [comment]
<p>We can rename the check to "ImageDatasetDrift" (I'm serious)</p>

---

 <a href='https://app.reviewnb.com/deepchecks/deepchecks/pull/879/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

File: deepchecks/vision/checks/distribution/whole_dataset_drift.py (hunk 4 above)
Comment (nirhutnik) [reply deepchecks/vision/checks/distribution/whole_dataset_drift.py:45]: I get it. 
But my reasons were:
1. If we want caching in the future, we need these functions as part of a class. And it makes sense that if a user passed alternative properties here, they'll do that in the other checks that use those as well.
2. In the case of passing callables, they need to pass both callable and str (func name). This makes for a really weird structure, of either receiving a list of str or a dict of str: callable. 

I agree my solution is not perfect, but not sure what I should do.

File: deepchecks/vision/checks/distribution/whole_dataset_drift.py (hunk 3 above)
Comment (noamzbr) [reply deepchecks/vision/checks/distribution/whole_dataset_drift.py:34]: You may say I'm commenting on stuff you didn't catch when reviewing my text 😉 
Anyway, yeah it suddenly bothered me, but if we're not doing it in other checks then no sense starting here.

**Review by noamzbr — APPROVED** [review]

**Pull request merged**